code/system/application/perception/src/detectlane/detectlane.cpp

```diff
@@ -94,437 +130,165 @@ void DetectLane::tearDown()
-  //   cv::Point pt2(cvRound(x0 - 2000*(-b)),
-  //              cvRound(y0 - 2000*(a)));
-
-  //   cv::line(intenImg, pt1, pt2, cv::Scalar(255,255,255), 1, 1 );
-  // }
-
-
-  // // Input Quadilateral or Image plane coordinates
-  // cv::Point2f inputQuad[4] = {cv::Point2f(350,470),cv::Point2f(750,470),cv::Point2f(2180,720),cv::Point2f(-700,720)};
-  // // Output Quadilateral or World plane coordinates
-  // cv::Point2f outputQuad[4] = {cv::Point2f(0,0),cv::Point2f(1080,0),cv::Point2f(1080,720),cv::Point2f(0,720)}; 
-  // // std::cout << inputQuad << outputQuad << std::endl;
-  // cv::Mat lambda = cv::getPerspectiveTransform(inputQuad, outputQuad);
-  // cv::Mat warpedImg = m_image.clone();
-  // cv::warpPerspective(m_image, warpedImg, lambda, warpedImg.size());
-  // cv::Point polypoint[4];
-  // for(int i=0;i<4;i++){
-  //     polypoint[i]= inputQuad[i];
-  // }
-  // const cv::Point* countours[1]={polypoint};
-  // int v = 4;
-  // const int* constCont = &v; 
-  // cv::polylines(m_image, countours, constCont, 4, 1, cv::Scalar(0,255,0), 4, 8, 0);
-
-
-  // Get parametric line representation
-  std::vector<cv::Vec2f> p, m;
-
-  GetParametricRepresentation(p, m, detectedLines);
-
-  // Get points on lines
-  std::vector<cv::Vec2f> xScreenP,yScreenP;
-  std::vector<cv::Vec2f> xWorldP, yWorldP;
-  GetPointsOnLine(xScreenP, yScreenP, xWorldP, yWorldP, p, m);
-
-  // Pair up lines to form a surface
-  std::vector<int8_t> groupIds;
-  GetLinePairs(xScreenP, groupIds);
-  if(groupIds.size() == 0){
+void DetectLane::UpdateVisualMemory() {
+  odcore::data::TimeStamp now;
+  cv::Rect rectROI(m_roi[0], m_roi[1], m_roi[2], m_roi[3]);
+  cv::Mat visualImpression;
+  try {
+    odcore::base::Lock l(m_mtx);
+    visualImpression = m_currentImg(rectROI).clone();
+  } catch (cv::Exception& e) {
+    std::cerr << "[Detect lane] Error cropping the image due to dimension size. " << std::endl;
     return;
   }
-
-  // std::cout << groupIds.size() << std::endl;
-
-
-  if(m_debug) {
-    for(uint8_t i = 0; i < groupIds.size(); i++) {
-      for(uint8_t k = 0; k < 2; k++) {
-        cv::circle(m_image, cv::Point(xScreenP[groupIds[i]][k],yScreenP[groupIds[i]][k]), 10, cv::Scalar(0,255,0), 3, 8);
-        // std::cout << xWorldP[groupIds[i]][k] << "x" << yWorldP[groupIds[i]][k] << ",";
-      }
-      // std::cout << std::endl;
-    }
-    cv::resize(cannyImg, display[0], cv::Size(windowWidth, windowHeight), 0, 0,
-      cv::INTER_CUBIC);
-    cv::imshow("FOE", display[0]);
-    cv::resize(m_image, display[1], cv::Size(windowWidth*2, windowHeight*2), 0, 0,
-      cv::INTER_CUBIC);
-    cv::imshow("FOE2", display[1]);
-    cv::waitKey(1);
+  
+  // Reduce the noise in image
+  cv::medianBlur(visualImpression, visualImpression, m_blurKernelSize);
+  
+  m_visualMemory.push_back(std::make_pair(now, visualImpression));
+  // Delete old mem
+  while(m_visualMemory.size() > 0 && (now-m_visualMemory[0].first).toMicroseconds()/1000000.0 > m_memThreshold){
```

Comment:
Missing spaces. Also, it's not nice to put so much action into a parenthesis, introduce a boolean.